File: src/test/java/com/integreety/yatspec/e2e/integration/matcher/InterceptedInteractionMatcher.java
package com.integreety.yatspec.e2e.integration.matcher;

import com.integreety.yatspec.e2e.captor.repository.model.InterceptedInteraction;
import com.integreety.yatspec.e2e.captor.repository.model.Type;
import lombok.RequiredArgsConstructor;
import org.hamcrest.Description;
import org.hamcrest.TypeSafeMatcher;

@RequiredArgsConstructor
public class InterceptedInteractionMatcher extends TypeSafeMatcher<InterceptedInteraction> {

    private final Type type;
    private final String serviceName;
    private final String body;
    private final String target;
    private final String path;

    @Override
    protected boolean matchesSafely(final InterceptedInteraction interceptedInteraction) {
        return interceptedInteraction.getType().equals(type)
                && interceptedInteraction.getServiceName().equals(serviceName)
                && interceptedInteraction.getBody().equals(body)
                && interceptedInteraction.getTarget().equals(target)
                && interceptedInteraction.getPath().equals(path);
    }

    @Override
    public void describeTo(final Description description) {
        description.appendText("an item with body:" + body + " and serviceName:" + serviceName + " and type:" + type + " and target:" + target + " and path:" + path);
    }

    public static InterceptedInteractionMatcher with(final Type type, final String serviceName, final String body, final String target, final String path) {
        return new InterceptedInteractionMatcher(type, serviceName, body, target, path);
    }
}

File: feign/src/integrationMongoTest/kotlin/io/lsdconsulting/lsd/distributed/interceptor/integration/matcher/InterceptedInteractionMatcher.kt
package io.lsdconsulting.lsd.distributed.interceptor.integration.matcher

import io.lsdconsulting.lsd.distributed.connector.model.InteractionType
import io.lsdconsulting.lsd.distributed.connector.model.InterceptedInteraction
import org.hamcrest.Description
import org.hamcrest.TypeSafeMatcher

class InterceptedInteractionMatcher(
    private val interactionType: InteractionType,
    private val serviceName: String,
    private val body: String?,
    private val target: String,
    private val path: String,
    private val httpStatus: String?,
    private val httpMethod: String?,
) : TypeSafeMatcher<InterceptedInteraction>() {

    override fun matchesSafely(interceptedInteraction: InterceptedInteraction): Boolean {
        return interceptedInteraction.interactionType == interactionType &&
                interceptedInteraction.serviceName == serviceName &&
                interceptedInteraction.body == body &&
                interceptedInteraction.target == target &&
                interceptedInteraction.path == path &&
                interceptedInteraction.httpStatus == httpStatus &&
                interceptedInteraction.httpMethod == httpMethod
    }

    override fun describeTo(description: Description) {
        description.appendText("an item with body:$body and serviceName:$serviceName and type:$interactionType and target:$target and path:$path")
    }

    companion object {
        fun with(
            interactionType: InteractionType,
            serviceName: String,
            body: String?,
            target: String,
            path: String,
            httpStatus: String?,
            httpMethod: String?,
        ): InterceptedInteractionMatcher {
            return InterceptedInteractionMatcher(interactionType, serviceName, body, target, path, httpStatus, httpMethod)
        }
    }
}
